core/dblookupext/historyRepository.go

```diff
@@ -227,57 +216,34 @@ func (hr *historyRepository) RegisterToBlockTracker(blockTracker BlockTracker) {
 		return
 	}
 
-	blockTracker.RegisterCrossNotarizedHeadersHandler(hr.onCrossNotarizedHeaders)
-	blockTracker.RegisterSelfNotarizedHeadersHandler(hr.onSelfNotarizedHeaders)
-}
-
-func (hr *historyRepository) onCrossNotarizedHeaders(shardID uint32, headers []data.HeaderHandler, headersHashes [][]byte) {
-	log.Debug("onCrossNotarizedHeaders()", "shardID", shardID)
-	hr.onNotarizedBlocks(shardID, headers, headersHashes)
-}
-
-func (hr *historyRepository) onSelfNotarizedHeaders(shardID uint32, headers []data.HeaderHandler, headersHashes [][]byte) {
-	log.Debug("onSelfNotarizedHeaders()", "shardID", shardID)
-	hr.onNotarizedBlocks(shardID, headers, headersHashes)
+	blockTracker.RegisterFinalMetachainHeadersHandler(hr.onNotarizedBlocks)
 }
 
-// This is called both with "shardID" == "metachain" (regular notarization events),
-// but also with "shardID" IN [0, 1, 2, ...] when blocks produced by the metachain (as a source) are notarized.
 // TODO: In the future, find a way to skip already processed notifications (e.g. using "headerHash" as a deduplication key).
 func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.HeaderHandler, headersHashes [][]byte) {
 	for i, headerHandler := range headers {
 		headerHash := headersHashes[i]
 
-		log.Debug("onNotarizedBlocks() on header:", "shardID", shardID, "headerHash", headerHash, "type", fmt.Sprintf("%T", headerHandler))
+		log.Debug("onNotarizedBlocks():", "shardID", shardID, "headerHash", headerHash, "type", fmt.Sprintf("%T", headerHandler))
 
 		metaBlock, isMetaBlock := headerHandler.(*block.MetaBlock)
 		if isMetaBlock {
 			for _, shardData := range metaBlock.ShardInfo {
 				hr.onNotarizedInMetaBlock(metaBlock.GetNonce(), headerHash, &shardData)
 			}
-			continue
+		} else {
+			log.Error("onNotarizedBlocks(): unexpected type of header", "type", fmt.Sprintf("%T", headerHandler))
 		}
-
-		// Fix this! this is imprecise:
-		header, isHeader := headerHandler.(*block.Header)
-		if isHeader {
-			hr.onNotarizedInRegularBlockOfMeta(header.GetNonce(), headerHash, header)
-			continue
-		}
-
-		log.Error("onNotarizedBlocks(): unexpected type of header", "type", fmt.Sprintf("%T", headerHandler))
 	}
 }
 
 func (hr *historyRepository) onNotarizedInMetaBlock(metaBlockNonce uint64, metaBlockHash []byte, shardData *block.ShardData) {
-	for _, miniblockHeader := range shardData.GetShardMiniBlockHeaders() {
-		hr.onNotarizedMiniblock(metaBlockNonce, metaBlockHash, shardData.GetShardID(), miniblockHeader)
+	if metaBlockNonce <= 1 {
```

Comment:
< 1 instead <= 1. If meta misses some rounds it can include in nonce 1 a lot of shardData if the shards worked good

Reply:
Fixed.